File: CircularArrayLoop2.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    bool circularArrayLoop(vector<int>& nums) {
    	int n = nums.size();
        for(int i = 0; i < n; i++) {
        	if(nums[i] == 0) continue;
        	int cur = nums[i];
        	nums[i] = 0;
        	int next = ((i + cur) % n + n) % n;
        	if(next == i) continue;
        	while(cur * nums[next] > 0) {
        		cur = nums[next];
        		nums[next] = 0;
        		next = ((next + cur) % n + n) % n;
			}
			if(next == i) return true;
		}
		return false;
    }
};

int main() {
	
	int n;
	cin>>n;
	
	vector<int> nums;
	for(int i = 0; i < n; i++) {
		int num;
		cin>>num;
		nums.push_back(num);
	}
	
	Solution *solution = new Solution();
	solution->circularArrayLoop(nums) ? cout<<"true" : cout<<"false";
	
	return 0;
	
}

File: CircularArrayLoop2.py
class Solution:
    def circularArrayLoop(self, nums) -> bool:
        def getIndex(i):
            return ((i + nums[i]) % n + n) % n

        n = len(nums)
        for i in range(n):
            if nums[i] == 0:
                continue
            j = i
            k = getIndex(i)
            while nums[i] * nums[k] > 0 and nums[getIndex(k)] * nums[i] > 0:
                if j == k:
                    if j == getIndex(j):
                        break
                    return True
                j = getIndex(j)
                k = getIndex(getIndex(k))

            j = i
            next = getIndex(j)
            while nums[j] * nums[next] > 0:
                next = getIndex(j)
                nums[j] = 0
                j = next
        return False


if __name__ == '__main__':
    nums = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.circularArrayLoop(nums))
